File: solve.java
/*
 * BENCH_BUILD_CMD:javac solve.java
 * BENCH_INVOKE_CMD:java solve ./dict.txt
 * BENCH_VERSION_CMD:java -version 2>&1 | awk '{print $3;exit}'
 */

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.InputStreamReader;

class Node {
	public boolean _finished = false;
	private HashMap<Character, Node> _next = new HashMap<Character, Node>();
	public void addSuffix( String suffix_ ) {
		if ( suffix_.isEmpty() ) {
			_finished = true;
		} else {
			char c = suffix_.charAt( 0 );
			Node n = _next.get( c );
			if ( n == null ) {
				n = new Node();
				_next.put( c, n );
			}
			n.addSuffix( suffix_.substring( 1 ) );
		}
	}

	Node get( char c_ ) {
		return ( _next.get( c_ ) );
	}

	void printAll( String prefix_ ) {
		if ( _finished ) {
			System.out.println( prefix_ );
		}
		for ( Map.Entry<Character, Node> n : _next.entrySet() ) {
			n.getValue().printAll( prefix_ + n.getKey() );
		}
	}
}

class Cube {
	public char _letter = 0;
	public boolean _visited = false;
	public ArrayList<Cube> _neighbors = new ArrayList<Cube>();
	Cube( char letter_ ) {
		_letter = letter_;
	}
}

class Board {
	private ArrayList<Cube> _cubes = new ArrayList<Cube>();
	private int _size = 0;
	public Board( String letters_ ) {
		int len = (int)Math.sqrt( (float)letters_.length() );
		assert( len * len == letters_.length() );
		_size = len;
		for ( char l : letters_.toCharArray() ) {
			_cubes.add( new Cube( l ) );
		}
		int[][] deltas = new int[][]{ { -1,-1 }, { -1,0 }, { -1,1 }, { 0,-1 }, { 0,1 }, { 1,-1 }, { 1,0 }, { 1,1 } };
		for ( int x = 0; x < len; ++ x ) {
			for ( int y = 0; y < len; ++ y ) {
				for ( int[] d : deltas ) {
					int nx = x + d[0];
					int ny = y + d[1];
					if ( ( nx >= 0 ) && ( nx < len ) && ( ny >= 0 ) && ( ny < len ) ) {
						get_cube( x, y )._neighbors.add( get_cube( nx, ny ) );
					}
				}
			}
		}
	}

	public Cube get_cube( int x_, int y_ ) {
		return ( _cubes.get( y_ * _size + x_ ) );
	}

	public ArrayList<String> solve( Node dictionary_ ) {
		HashSet<String> result = new HashSet<String>();
		for ( Cube cube : _cubes ) {
			solveRecursive( result, "", cube, dictionary_ );
		}
		ArrayList<String> resultSorted = new ArrayList<String>( result );
		java.util.Collections.sort(
			resultSorted,
			new java.util.Comparator<String>() {
				public int compare(String s1, String s2) {
					return s1.length() - s2.length();
		    }
			}
		);
		return ( resultSorted );
	}

	void solveRecursive( HashSet<String> result, String prefix_, Cube cube_, Node dictNode_ ) {
		Node nextNode = dictNode_.get( cube_._letter );
		if ( nextNode == null ) {
			return;
		}
		cube_._visited = true;
		String newPrefix = prefix_ + cube_._letter;
		if ( nextNode._finished && ( newPrefix.length() >= 3 ) ) {
			result.add( newPrefix );
		}
		for ( Cube neighbor : cube_._neighbors ) {
			if ( ! neighbor._visited ) {
				solveRecursive( result, newPrefix, neighbor, nextNode );
			}
		}
		cube_._visited = false;
	}
}

public class solve {
	void doSolve( String file_ ) throws java.io.FileNotFoundException, java.io.IOException {
		System.out.println( "[>] Loading dictionary from file: " + file_ );
		Node dictionaryRoot = new Node();
		try ( BufferedReader br = new BufferedReader( new FileReader( file_) ) ) {
			String line;
			while ( ( line = br.readLine() ) != null ) {
				dictionaryRoot.addSuffix( line );
			}
		}
//		dictionaryRoot.printAll( "" );
		System.out.println( "[ OK ] Ready" );
		BufferedReader br = new BufferedReader( new InputStreamReader( System.in ) );
		String letters = "";
		String line = null;
		while ( ( line = br.readLine() ) != null ) {
			line = line.trim();
			if ( ( letters.length() > 0 ) && ( line.length() == 0 ) ) {
				System.out.println( "[ OK ] Solving" );
				Board board = new Board( letters );
				for ( String word : board.solve( dictionaryRoot ) ) {
					System.out.println( "(" + word.length() + ") " + word );
				}
				System.out.println( "[ OK ] Solved" );
				letters = "";
			} else {
				letters = letters + line;
			}
		}
		return;
	}
	public static void main( String[] argv_ ) {
		System.out.println( "[>] Boggle solver" );
		try {
			solve s = new solve();
			if ( argv_.length != 1 ) {
				throw new RuntimeException( "Specify one argument which should be path to dictionary file." );
			}
			s.doSolve( argv_[0] );
		} catch ( Exception e ) {
			e.printStackTrace();
			System.exit( 1 );
		}
		return;
	}
}

/* vim: set ft=java: */

File: solve.go
/*
 * BENCH_BUILD_CMD:go build solve.go
 * BENCH_INVOKE_CMD:./solve ./dict.txt
 * BENCH_VERSION_CMD:go version | awk '{print $3}'
 */
package main

import "fmt"
import "os"
import "bufio"
import "math"
import "sort"
import "strings"

type DictionaryNode struct {
	_finished bool
	_next map[byte]*DictionaryNode
}

func NewDictionaryNode() *DictionaryNode {
	n := new( DictionaryNode )
	n._next = make( map[byte]*DictionaryNode )
	return ( n )
}

func ( node_ *DictionaryNode ) add_suffix( suffix_ string ) {
	if ( len( suffix_ ) == 0 ) {
		node_._finished = true
	} else {
		char := suffix_[0]
		_, hasTail := node_._next[char]
		if ( ! hasTail ) {
			node_._next[char] = NewDictionaryNode()
		}
		node_._next[char].add_suffix( suffix_[1:] )
	}
}

func ( node_ *DictionaryNode ) get( char_ byte ) *DictionaryNode {
	node, hasTail := node_._next[char_]
	if ( ! hasTail ) {
		node = nil
	}
	return ( node )
}

func ( node_ *DictionaryNode ) print_all( prefix_ string ) {
	if ( node_._finished ) {
		fmt.Printf( prefix_ + "\n" )
	}
	for char, tail := range node_._next {
		tail.print_all( prefix_ + string( char ) )
	}
}

type Cube struct {
	_letter byte
	_visited bool
	_neighbors []*Cube
}

func NewCube( letter_ byte ) *Cube {
	c := new( Cube )
	c._letter = letter_
	c._neighbors = make( []*Cube, 0 )
	return ( c )
}

type Board struct {
	_cubes []*Cube
	_size int
}

func NewBoard( letters_ string ) *Board {
	b := new( Board )
	size := int( math.Sqrt( float64( len( letters_ ) ) ) )
	if ( size * size != len( letters_ ) ) {
		return ( nil )
	}
	b._size = size
	b._cubes = make( []*Cube, 0 )
	for i := range letters_ {
		b._cubes = append( b._cubes, NewCube( letters_[i] ) )
	}

	deltas := [8][2]int{{-1, -1}, {-1, 0}, {-1, 1}, {0, -1}, {0, 1}, {1, -1}, {1, 0}, {1, 1}}
	for x := 0; x < size; x ++ {
		for y := 0; y < size; y ++ {
			for _, d := range deltas {
				nx := x + d[0]
				ny := y + d[1]
				if ( ( nx >= 0 ) && ( nx < size ) && ( ny >= 0 ) && ( ny < size ) ) {
					b.get_cube( x, y )._neighbors = append( b.get_cube( x, y )._neighbors, b.get_cube( nx, ny ) )
				}
			}
		}
	}
	return ( b )
}

func ( board_ *Board ) get_cube( x_ int, y_ int ) *Cube {
	return ( board_._cubes[y_ * board_._size + x_] );
}

type ByLength []string
func ( a ByLength ) Len() int {
	return ( len ( a ) )
}
func ( a ByLength ) Swap( i int, j int ) {
	a[i], a[j] = a[j], a[i]
}
func ( a ByLength ) Less( i int, j int ) bool {
	return ( len( a[i] ) < len( a[j] ) )
}

func ( board_ *Board ) solve( dictionary_ *DictionaryNode ) []string {
	result := make(map[string]bool);
	for _, cube := range board_._cubes {
		board_.solve_recursive( result, "", cube, dictionary_ );
	}
	var sorted []string
	for k := range result {
		sorted = append( sorted, k )
	}
	sort.Sort( ByLength( sorted ) )
	return ( sorted )
}

func ( board_ *Board ) solve_recursive( result map[string]bool, prefix_ string, cube_ *Cube, dictNode_ *DictionaryNode ) {
	nextNode := dictNode_.get( cube_._letter );
	if ( nextNode == nil ) {
		return;
	}
	cube_._visited = true;
	newPrefix := prefix_ + string( cube_._letter );
	if ( nextNode._finished && ( len( newPrefix ) >= 3 ) ) {
		result[newPrefix] = true
	}
	for _, neighbor := range cube_._neighbors {
		if ( ! neighbor._visited ) {
			board_.solve_recursive( result, newPrefix, neighbor, nextNode );
		}
	}
	cube_._visited = false;
}

func main() {
	file, _ := os.Open( os.Args[1] )
	defer file.Close()

	scanner := bufio.NewScanner( file )
	dictionaryRoot := NewDictionaryNode()
	for scanner.Scan() {
		dictionaryRoot.add_suffix( scanner.Text() )
	}
//	dictionaryRoot.print_all( "" )

	fmt.Printf( "[ OK ] Ready\n" )

	input := bufio.NewReader( os.Stdin )
	letters := ""
	for {
		line, _ := input.ReadString( '\n' )
		if ( len( line ) == 0 ) {
			break;
		}
		line = strings.TrimSpace( line )
		if ( ( len( letters ) > 0 ) && ( len( line ) == 0 ) ) {
			fmt.Printf( "[ OK ] Solving\n" )
			board := NewBoard( letters )
			if ( board != nil ) {
				for _, word := range board.solve( dictionaryRoot ) {
					fmt.Printf( "(%d) %s\n", len( word ), word )
				}
				fmt.Printf( "[ OK ] Solved\n" )
			} else {
				fmt.Printf( "Bad board definition!\n" )
			}
			letters = ""
		} else {
			letters = letters + line
		}
	}
	return
}

/* vim: set ft=go: */
